rewrite-java/src/main/java/org/openrewrite/java/cleanup/ReplaceRedundantFormatWithPrintf.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.List;
+
+public class ReplaceRedundantFormatWithPrintf extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Replace redundant String format invocations that are wrapped with PrintStream operations";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replaces `PrintStream.print(String.format(format, ...args))` with `PrintStream.printf(format, ...args)` (and for `println`, appends a newline to the format string).";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                method = super.visitMethodInvocation(method, ctx);
+
+                if (!isPrintStreamPrintOrPrintln(method.getMethodType())) {
+                    return method;
+                }
+
+                if (method.getArguments().size() != 1) {
+                    return method;
+                }
+
+                Expression arg = method.getArguments().get(0);
+                if (!(arg instanceof J.MethodInvocation)) {
+                    return method;
+                }
+
+                J.MethodInvocation innerMethodInvocation = (J.MethodInvocation) arg;
+
+                if (!isStringFormat(innerMethodInvocation.getMethodType())) {
+                    return method;
+                }
+
+                List<Expression> newArguments = innerMethodInvocation.getArguments();
+
+                boolean needsNewline = method.getMethodType().getName().equals("println");
```

Comment:
You can use `MethodMatcher` to match methods precisely:
```java
MethodMatcher m = new MethodMatcher("java.io.PrintStream println(..)");
...

if(m.matches(method)) { 
   ...
}
```

https://public.moderne.io/recipes/org.openrewrite.java.search.FindMethods?defaults=W3sibmFtZSI6Im1ldGhvZFBhdHRlcm4iLCJ2YWx1ZSI6Im9yZy5vcGVucmV3cml0ZS5qYXZhLk1ldGhvZE1hdGNoZXIgPGNvbnN0cnVjdG9yPiguLikifSx7Im5hbWUiOiJmbG93IiwidmFsdWUiOiJkYXRhIn1d
See

Reply:
Awesome, thank you!